hook alert / confirm / prompt dialog and implement message select dialog

yukinkomob opened the pull request:
### 実装

* alert / confirm / prompt ダイアログのフックを実装
* ダイアログ表示にダークテーマを設定
* メッセージ選択ダイアログを実装
    * prompt ダイアログの message の値が "メッセージ文字列を入力して下さい。" の場合に、メッセージ選択ダイアログを表示
    * デフォルトのメッセージ文字列リストを表示
    * CLEAR ボタンを押下すると EditText のメッセージがクリアされる
    * OK ボタンを押下すると EditText の文字列を Web に受け渡す
    * Cancel ボタンを押下するとダイアログを閉じる

### 表示

* Alert ダイアログ

![Screen Shot 2022-06-24 at 3 56 28](https://user-images.githubusercontent.com/3205581/175377386-f1b26ea7-baea-431f-bcf7-89e517195924.png)

* メッセージ選択ダイアログ

![Screen Shot 2022-06-24 at 4 01 18](https://user-images.githubusercontent.com/3205581/175377410-b2484f59-db60-49a7-8566-ec3222597a3e.png)

yukinkomob committed:
import the material dialog library

yukinkomob committed:
change to that MyWebChromeClient uses AppCompatActivity instead of MainActivity

yukinkomob committed:
implement alert, confirm and prompt

yukinkomob committed:
set hint string on prompt

yukinkomob committed:
set theme for material dialog

yukinkomob committed:
implement message select dialog

katahiromz commented on app/src/main/java/com/katahiromz/krakra_ja_jp/MyWebChromeClient.kt:
```diff
@@ -32,26 +44,129 @@ class MyWebChromeClient(activity: MainActivity) : WebChromeClient() {
+            message(text = message)
+            positiveButton(text = "OK") { }
+        }
+        return true
```
`result?.confirm()`が必要。

yukinkomob replied:
対応しました。

katahiromz commented on app/src/main/java/com/katahiromz/krakra_ja_jp/MyWebChromeClient.kt:
```diff
@@ -32,26 +44,129 @@ class MyWebChromeClient(activity: MainActivity) : WebChromeClient() {
+            input(hint = activity.getString(R.string.prompt_hint)) { _, text ->
+                inputtedText = text.toString()
+            }
+            positiveButton(text = "OK") {
```
リソース文字列を使う。

yukinkomob replied:
対応しました。

katahiromz commented on app/src/main/java/com/katahiromz/krakra_ja_jp/MyWebChromeClient.kt:
```diff
@@ -32,26 +44,129 @@ class MyWebChromeClient(activity: MainActivity) : WebChromeClient() {
+            positiveButton(text = "OK") {
+                result?.confirm(inputtedText ?: "")
+            }
+            negativeButton(text = "Cancel") {
```
リソース文字列を使う。

yukinkomob replied:
対応しました。

katahiromz commented on app/src/main/java/com/katahiromz/krakra_ja_jp/MyWebChromeClient.kt:
```diff
@@ -32,26 +44,129 @@ class MyWebChromeClient(activity: MainActivity) : WebChromeClient() {
+        return true
+    }
+
+    private fun isSelectMessageDialog(message: String?): Boolean = message == "メッセージ文字列を入力して下さい。"
```
リソース文字列を使う。

yukinkomob replied:
対応しました。

katahiromz commented on app/src/main/java/com/katahiromz/krakra_ja_jp/MyWebChromeClient.kt:
```diff
@@ -32,26 +44,129 @@ class MyWebChromeClient(activity: MainActivity) : WebChromeClient() {
+            title(text = title)
+            message(text = message)
+            customView(R.layout.message_select_dialog, scrollable = true, horizontalPadding = true)
+            positiveButton(text = "OK") {
```
リソース文字列を使う。

yukinkomob replied:
対応しました。

katahiromz commented on app/src/main/java/com/katahiromz/krakra_ja_jp/MyWebChromeClient.kt:
```diff
@@ -32,26 +44,129 @@ class MyWebChromeClient(activity: MainActivity) : WebChromeClient() {
+                val editText = getCustomView().findViewById<EditText>(R.id.message_edit)
+                result?.confirm(editText.text.toString())
+            }
+            negativeButton(text = "Cancel") {
```
リソース文字列を使う。

yukinkomob replied:
対応しました。

katahiromz commented on app/src/main/java/com/katahiromz/krakra_ja_jp/MyWebChromeClient.kt:
```diff
@@ -32,26 +44,129 @@ class MyWebChromeClient(activity: MainActivity) : WebChromeClient() {
+        MaterialDialog(activity).show {
+            title(text = title)
+            message(text = message)
+            positiveButton(text = "OK") { }
```
リソース文字列を使う。

yukinkomob replied:
対応しました。

katahiromz commented on app/src/main/java/com/katahiromz/krakra_ja_jp/MyWebChromeClient.kt:
```diff
@@ -32,26 +44,129 @@ class MyWebChromeClient(activity: MainActivity) : WebChromeClient() {
+        MaterialDialog(activity).show {
+            title(text = title)
+            message(text = message)
+            positiveButton(text = "OK") {
```
リソース文字列を使う。

yukinkomob replied:
対応しました。

katahiromz commented on app/src/main/java/com/katahiromz/krakra_ja_jp/MyWebChromeClient.kt:
```diff
@@ -32,26 +44,129 @@ class MyWebChromeClient(activity: MainActivity) : WebChromeClient() {
+            positiveButton(text = "OK") {
+                result?.confirm()
+            }
+            negativeButton(text = "Cancel") {
```
リソース文字列を使う。

yukinkomob replied:
対応しました。

katahiromz commented on app/src/main/res/layout/message_select_dialog.xml:
```diff
@@ -0,0 +1,44 @@
+        android:layout_height="wrap_content"
+        android:layout_marginStart="8dp"
+        android:layout_marginEnd="8dp"
+        android:text="clear"
```
リソース文字列を使う。

katahiromz replied:
視覚障害者のために`android:contentDescription`を追加して下さい。

yukinkomob replied:
対応しました。

katahiromz commented on app/src/main/res/values/themes.xml:
```diff
@@ -5,13 +5,19 @@
         <item name="colorSecondary">@color/themeColorDark</item>
         <item name="colorSecondaryVariant">@color/themeColor</item>
         <item name="colorOnSecondary">@color/black</item>
+        <item name="android:textColorSecondary">@color/white</item>
```
CLEARの文字色は黒がいいかな。

yukinkomob replied:
対応しました。
上記の指定箇所はボタン文字色以外にも影響する箇所なので、ボタン文字色はレイアウトファイルにて指定しました。

katahiromz commented on app/src/main/java/com/katahiromz/krakra_ja_jp/MyWebChromeClient.kt:
```diff
@@ -32,26 +44,129 @@ class MyWebChromeClient(activity: MainActivity) : WebChromeClient() {
+        val title = activity.getString(R.string.app_name)
+        if (isSelectMessageDialog(message)) {
+            // メッセージ選択ダイアログを表示
+            showSelectMessageDialog(title = title, message = message, result = result)
```
`defaultValue`を使って下さい。

yukinkomob replied:
対応しました。

katahiromz commented on app/src/main/java/com/katahiromz/krakra_ja_jp/MyWebChromeClient.kt:
```diff
@@ -32,26 +44,129 @@ class MyWebChromeClient(activity: MainActivity) : WebChromeClient() {
+        var totalHeight = 0
+
+        // 個々のアイテムの高さを測り、加算していく
+        for (i in 0 until arrayAdapter.count)
```
```kotlin
        for (i in 0 until arrayAdapter.count) {
```

yukinkomob replied:
対応しました。

katahiromz commented on app/src/main/java/com/katahiromz/krakra_ja_jp/MyWebChromeClient.kt:
```diff
@@ -32,26 +44,129 @@ class MyWebChromeClient(activity: MainActivity) : WebChromeClient() {
+        var inputtedText: String? = null
+        MaterialDialog(activity).show {
+            title(text = title)
+            message(text = message)
```
`defaultValue`を使って下さい。

katahiromz commented on app/src/main/java/com/katahiromz/krakra_ja_jp/MyWebChromeClient.kt:
```diff
@@ -32,26 +44,129 @@ class MyWebChromeClient(activity: MainActivity) : WebChromeClient() {
+        MaterialDialog(activity).show {
+            title(text = title)
+            message(text = message)
+            input(hint = activity.getString(R.string.prompt_hint)) { _, text ->
```
`defaultValue`を使って下さい。

yukinkomob replied:
対応しました。

katahiromz commented on app/src/main/res/layout/message_select_dialog.xml:
```diff
@@ -0,0 +1,44 @@
+        android:layout_width="match_parent"
+        android:layout_height="250dp"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
```
視覚障害者のために`android:contentDescription`を追加して下さい。

yukinkomob replied:
対応しました。

katahiromz commented on app/src/main/res/layout/message_select_dialog.xml:
```diff
@@ -0,0 +1,44 @@
+        app:layout_constraintEnd_toStartOf="@id/clear_button"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@id/message_list"
+        tools:ignore="Autofill,HardcodedText,LabelFor"
```
視覚障害者のために`android:contentDescription`を追加して下さい。

yukinkomob replied:
ここは hint を指定済みなので、contentDescription は不要のようです。

![Screen Shot 2022-06-24 at 22 06 53](https://user-images.githubusercontent.com/3205581/175542951-b79833fa-d7c2-4c78-864c-d4fa89885de8.png)

katahiromz commented:
`values-night`は消してもいいかも。

katahiromz commented on app/src/main/java/com/katahiromz/krakra_ja_jp/MyWebChromeClient.kt:
```diff
@@ -32,26 +44,129 @@ class MyWebChromeClient(activity: MainActivity) : WebChromeClient() {
+            negativeButton(text = "Cancel") {
+                result?.cancel()
+            }
+            lifecycleOwner(activity)
```
このライフサイクルは何のためか説明できますか？

yukinkomob replied:
lifecycleOwner の説明を見ると下記のようにありますので、ご参照ください。

```
Attach the dialog to a lifecycle and dismiss it when the lifecycle is destroyed. Uses the given owner lifecycle if provided, else falls back to the Context of the dialog window if it can.
Params:
owner --Optional lifecycle owner, if its null use windowContext.
```

要はダイアログインスタンスのメモリ管理を、指定したライフサイクルオーナーと同期するための宣言であり、今回は activity をライフサイクルオーナーとして指定しているため、activity が破棄されるのに合わせてダイアログインスタンスのメモリも破棄されることが保証されます。

ここ以外のダイアログの処理にもライフサイクルの指定を追加いたしました。

yukinkomob committed:
delete night theme

yukinkomob committed:
corrected according to the review indication
- replacement with string resource
- set content description
- set the defaultValue of the dialog as the initial value of the input box
- disables the operation to cancel when tapping outside the dialog
- set the text color of the clear button to black
- specify life cycle owner in dialog

katahiromz commented on app/src/main/res/values/themes.xml:
```diff
@@ -5,13 +5,19 @@
         <item name="colorPrimary">@color/themeColor</item>
         <item name="colorPrimaryVariant">@color/themeColorDark</item>
         <item name="colorOnPrimary">@color/white</item>
+        <item name="android:textColorPrimary">@color/themeColorDark</item>
```
```suggestion
        <item name="android:textColorPrimary">@color/themeColor</item>
```

katahiromz replied:
もう少し明るい色がいいということです。

yukinkomob replied:
変更しました。

katahiromz commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,5 +1,24 @@
+
+    <!-- message select dialog -->
+    <string name="message_select_dialog_message">メッセージ文字列を入力して下さい。</string>
+    <string name="message_clear_button">CLEAR</string>
```
```suggestion
    <string name="message_clear_button">クリア</string>
```

katahiromz replied:
日本語っぽく。

yukinkomob replied:
変更しました。

katahiromz commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,5 +1,24 @@
 <resources>
+    <!-- common -->
+    <string name="ok">OK</string>
+    <string name="cancel">Cancel</string>
```
```suggestion
    <string name="cancel">キャンセル</string>
```

katahiromz replied:
日本語っぽく。

yukinkomob replied:
変更しました。

katahiromz commented:
再テストお願いします。

yukinkomob committed:
set cancelable to disable back key on dialog

yukinkomob commented:
### テスト結果

- [x] 字ボタン押下時にメッセージ選択ダイアログを表示
- [x] メッセージ選択ダイアログにて、リスト項目をタップすると文字列が入力ボックスに表示される
- [x] メッセージ選択ダイアログにて、クリアボタンをタップすると入力ボックスの文字列が消去される
- [x] メッセージ選択ダイアログにて、OKボタンをタップすると入力ボックスの文字列が画面に表示される
- [x] メッセージ選択ダイアログにて、Cancelボタンをタップするとダイアログが閉じる
- [x] メッセージ選択ダイアログにて、バックキーを押下してもダイアログが閉じない
- [x] メッセージ選択ダイアログにて、ダイアログ外をタップしてもダイアログが閉じない
- [x] ヘルプボタンを押下して、アプリ初期化ボタンをタップするとフックしてアラートダイアログが開く

katahiromz commented:
上記の修正後、スクリーンショットもお願いします。

yukinkomob commented:
![Screen Shot 2022-06-25 at 0 09 44](https://user-images.githubusercontent.com/3205581/175564397-62f156fa-70da-45cd-a8bd-a62fce1add84.png)

yukinkomob commented:
![Screen Shot 2022-06-25 at 0 09 54](https://user-images.githubusercontent.com/3205581/175564463-1a04230c-be5c-471a-be6d-b1f63269360a.png)

yukinkomob commented:
@katahiromz 
修正しました。ご確認願います。

katahiromz commented:
「決定」は「OK」に戻して下さい。

yukinkomob committed:
update resources

yukinkomob commented:
> 「決定」は「OK」に戻して下さい。
戻しました。

katahiromz reviewed (approved):
いいですね。
